### PR TITLE
documentation: Secure boot quick start / walkthough

### PR DESCRIPTION
**Pull request by timg236**
Create a step by step walk through for enabling creating an EEPROM with signed boot enabled and booting a minimal OS image signed with that key.

This does not require changes to OTP so is safe for developers creating OS images to experiment with.

### CODE REVIEW

**Comment by timg236** [comment]
@peterharperuk

File: Readme.md [hunk 1]
```diff
@@ -91,6 +91,9 @@ Secure Boot requires the latest stable bootloader image.
 WARNING: If the `revoke_devkey` option is used to revoke the ROM development key then it will
 not be possible to downgrade to a bootloader older than 2022-01-06 OR disable secure-boot mode.
 
+### Tutorial
+Creating a secure-boot system from scratch be quite complex. The [secure-boot tutorial](secure-boot-example/README.md) uses a minimal example OS image to demonstrate how the Raspberry Pi specific aspects of secure-boot work.
```
Comment (peterharperuk) [inline Readme.md:95]: be quite complex -> CAN be quite complex

File: secure-boot-example/README.md [hunk 2]
```diff
@@ -1,8 +1,153 @@
-This directory contains an example secure boot image signed with the example private key in this directory.
+# Secure-boot quickstart
+
+## Overview
+Secure-boot is a mechanism for verifying the integrity of the kernel image and 
+other files required during boot by storing them in a signed disk image.
+These files include the GPU firmware (start.elf etc), Kernel, initrd, device-tree tree
+and overlays.
+Secure-boot does require any particular OS and does not extend into the OS
```
Comment (peterharperuk) [inline secure-boot-example/README.md:8]: does require -> does NOT? require

Comment (peterharperuk) [reply secure-boot-example/README.md:8]: "does not extend into the OS" I might suggest rewording this - does not require any support from the OS?

File: secure-boot-example/README.md [hunk 3]
```diff
@@ -1,8 +1,153 @@
...
+running on the CPU.  However, since the firmware will only load trusted 
+kernel + initrd images the kernel/initrd can safely verify or mount an
+an encrypted file-system e.g. by using a TPM.
+
+N.B The memory for the secure ramdisk is reclaimed as soon as the CPU is started.
```
Comment (peterharperuk) [inline secure-boot-example/README.md:13]: Maybe clarify that secure boot uses a ram disk?

File: secure-boot-example/README.md [hunk 4]
```diff
@@ -1,8 +1,153 @@
...
+
+## Example OS
+This example includes a very simple OS image for Compute Module 4 generated
```
Comment (peterharperuk) [inline secure-boot-example/README.md:16]: a very simple OS image for Compute Module 4 generated using buildroot -> reword?
a very simple OS image generated using buildroot for Compute Module 4

File: secure-boot-example/README.md [hunk 5]
```diff
@@ -1,8 +1,153 @@
...
+using buildroot in order to demonstrate how to sign and run OS images.
+
+The example does NOT modify the OTP or make other permenant changes to the system
```
Comment (peterharperuk) [inline secure-boot-example/README.md:19]: sp permenant -> permanent

File: secure-boot-example/README.md [hunk 6]
```diff
@@ -1,8 +1,153 @@
...
+and the code-signing can be disabled by reflashing the default bootloader EEPROM.
+
+The tutorial does not cover how to create the `boot.img` file or permenantly
+require secure-boot in OTP. Please see the top level [README](../Readme.md#building) and
+[secure-boot-recovery/README](../secure-boot-recovery/README.md] guides for
+more information.
+
+### Requirements for running this example
+* A Raspberry Pi Compute Module 4 which has not had secure-boot enabled in OTP.
```
Comment (peterharperuk) [inline secure-boot-example/README.md:28]: Maybe add a comment that CM4 ships wiithout secure boot enabled?

File: secure-boot-example/README.md [hunk 7]
```diff
@@ -1,8 +1,153 @@
...
+
+### Generate a signing key
+Secure-boot requires a 2048 bit RSA private key. You can either use a pre-existing
+key or generate an specific key for this example. The `KEY_FILE` environment variable
+used in the following instructions must contain the absolute path to the RSA private key in
+PEM format.
+```bash
+openssl genrsa 2048 > private.pem
+export KEY_FILE=$(pwd)/private.pem
+```
```
Comment (peterharperuk) [inline secure-boot-example/README.md:82]: I wonder if it' worth adding a note that for a real secure boot system, this should be kept private and secure?

File: secure-boot-example/README.md [hunk 8]
```diff
@@ -1,8 +1,153 @@
...
+```bash
+cd secure-boot-recovery
+# Generate the signed EEPROM image.
+../tools/update-pieeprom.sh -k "${KEY_FILE}"
+cd ..
+./rpiboot -d secure-boot-recovery
+```
+
+At this stage the signed image requirement can be reverted by flashing a default, unsigned image. 
+However, once the [OTP secure-boot flags](../secure-boot-recovery/README.md#locking-secure-boot-mode) are set then `SIGNED_BOOT` is permenantely enabled and cannot be overriden via the EEPROM config.
```
Comment (peterharperuk) [inline secure-boot-example/README.md:100]: I would clarify that we're not enabling secure boot in these instructions?

File: secure-boot-example/README.md [hunk 9]
```diff
@@ -1,8 +1,153 @@
...
+```bash
+openssl genrsa 2048 > private.pem
+export KEY_FILE=$(pwd)/private.pem
+```
+
+### Update the EEPROM to require signed OS images
+Enable `rpiboot` mode and flash the bootloader EEPROM with updated setting enables code-signing.
+
+The `boot.conf` file sets the `SIGNED_BOOT` property `1` which instructs the bootloader to only
+load files from `boot.img` and verify this against the signature in `boot.sig`. The `update-pieeprom.sh`
```
Comment (peterharperuk) [inline secure-boot-example/README.md:88]: I think it's worth listing what these files are,
e.g. boot.img and boot.sig are loaded by the bootloader to boot the operating system
pieeprom.bin, pieeprom.sig the bootloader image

File: secure-boot-example/README.md [hunk 10]
```diff
@@ -1,8 +1,153 @@
...
+### Mount the CM4 SD/EMMC after enabling secure-boot
+Now that `SIGNED_BOOT` is enabled the bootloader will only load images signed with private key generated earlier. 
+To boot the Compute Module in mass-storage mode a signed version of this code must be generated.  
+  
+**This signed-image should not be made available for download because it gives access to the EMMC as a block device**
+
+
+#### Sign the mass storage firmware image
+```bash
+cd secure-boot-msd
```
Comment (peterharperuk) [inline secure-boot-example/README.md:130]: I wonder if it's worth - near the top - explaining that the different sub folders are used OS images to boot the system via rpiboot for different purposes. As a newby it's maybe non-obvious what's going on here?

File: secure-boot-example/README.md [hunk 11]
```diff
@@ -1,8 +1,153 @@
...
+./rpiboot -d secure-boot-msd
 ```
 
-Clearly, product releases should never be signed with `example-private.pem`.
+### Loading `boot.img` from SD/EMMC.
+The bootloader can load a ramdisk `boot.img` from any of the bootable modes defined by the [BOOT_ORDER](https://www.raspberrypi.com/documentation/computers/raspberry-pi.html#BOOT_ORDER) EEPROM config setting.
+
+For example.
+
+* Boot the CM4 in MSD mode using as explained in the previous step.
```
Comment (peterharperuk) [inline secure-boot-example/README.md:146]: remove "using".

File: Readme.md (hunk 1 above)
Comment (timg236) [reply Readme.md:95]: done

File: secure-boot-example/README.md (hunk 3 above)
Comment (timg236) [reply secure-boot-example/README.md:13]: Updated the first sentence to sway "storing them in a secure ramdisk"

File: secure-boot-example/README.md (hunk 4 above)
Comment (timg236) [reply secure-boot-example/README.md:16]: done

File: secure-boot-example/README.md (hunk 5 above)
Comment (timg236) [reply secure-boot-example/README.md:19]: done

File: secure-boot-example/README.md (hunk 7 above)
Comment (timg236) [reply secure-boot-example/README.md:82]: done

File: secure-boot-example/README.md (hunk 8 above)
Comment (timg236) [reply secure-boot-example/README.md:100]: tweaked first sentence

File: secure-boot-example/README.md (hunk 10 above)
Comment (timg236) [reply secure-boot-example/README.md:130]: Added a link to top level README where this is documented

File: secure-boot-example/README.md (hunk 9 above)
Comment (timg236) [reply secure-boot-example/README.md:91]: There's quite of them kernel, overlays etc

File: secure-boot-example/README.md (hunk 6 above)
Comment (timg236) [reply secure-boot-example/README.md:28]: I'll remove the later part of the sentence, if someone has already enabled secure-boot then this is redundant

**Comment by peterharperuk** [comment]
Ok. Looks good to me

**Comment by timg236** [comment]
@JamesH65
